power_grid_model_c/power_grid_model/include/power_grid_model/main_core/output.hpp

```diff
@@ -98,351 +98,335 @@ constexpr ResIt produce_output(MainModelState<ComponentContainer> const& state,
-            case branch3_3:
-                return power_sensor.get_output<sym>(math_output[obj_math_id.group].branch[obj_math_id.pos].s_f);
-            case branch_to:
-                return power_sensor.get_output<sym>(math_output[obj_math_id.group].branch[obj_math_id.pos].s_t);
-            case source:
-                return power_sensor.get_output<sym>(math_output[obj_math_id.group].source[obj_math_id.pos].s);
-            case shunt:
-                return power_sensor.get_output<sym>(math_output[obj_math_id.group].shunt[obj_math_id.pos].s);
-            case load:
-            case generator:
-                return power_sensor.get_output<sym>(math_output[obj_math_id.group].load_gen[obj_math_id.pos].s);
-            case node:
-                return power_sensor.get_output<sym>(math_output[obj_math_id.group].bus_injection[obj_math_id.pos]);
-            default:
-                throw MissingCaseForEnumError(std::string(GenericPowerSensor::name) + " output_result()",
-                                              terminal_type);
-            }
-        });
+constexpr auto output_result(Component const& power_sensor, MainModelState<ComponentContainer> const& state,
+                             std::vector<MathOutputType> const& math_output, Idx const obj_seq) {
+    using sym = typename MathOutputType::sym;
+
+    auto const terminal_type = power_sensor.get_terminal_type();
+    Idx2D const obj_math_id = [&]() {
+        switch (terminal_type) {
+            using enum MeasuredTerminalType;
+
+        case branch_from:
+        case branch_to:
+            return state.topo_comp_coup->branch[obj_seq];
+        case source:
+            return state.topo_comp_coup->source[obj_seq];
+        case shunt:
+            return state.topo_comp_coup->shunt[obj_seq];
+        case load:
+        case generator:
+            return state.topo_comp_coup->load_gen[obj_seq];
+        // from branch3, get relevant math object branch based on the measured side
+        case branch3_1:
+            return Idx2D{state.topo_comp_coup->branch3[obj_seq].group, state.topo_comp_coup->branch3[obj_seq].pos[0]};
+        case branch3_2:
+            return Idx2D{state.topo_comp_coup->branch3[obj_seq].group, state.topo_comp_coup->branch3[obj_seq].pos[1]};
+        case branch3_3:
+            return Idx2D{state.topo_comp_coup->branch3[obj_seq].group, state.topo_comp_coup->branch3[obj_seq].pos[2]};
+        case node:
+            return state.topo_comp_coup->node[obj_seq];
+        default:
+            throw MissingCaseForEnumError(std::string(GenericPowerSensor::name) + " output_result()", terminal_type);
+        }
+    }();
+
+    if (obj_math_id.group == -1) {
+        return power_sensor.template get_null_output<sym>();
+    }
+
+    switch (terminal_type) {
+        using enum MeasuredTerminalType;
+
+    case branch_from:
+    // all power sensors in branch3 are at from side in the mathematical model
+    case branch3_1:
+    case branch3_2:
```

Comment:
No `[[fallthrough]];` here for shorter code? ;)

Reply:
`[[fallthrough]]` is **strongly** recommended if there is other code of the sort

```cpp
case foo:
  bar();
  [[fallthrough]];  // otherwise, bar() obscures the missing break statement
case baz:
  bla();
  break;
...
```

when there is no code in-between, there is no recommendation. i'm fine either way. if you prefer, I can add them but then I'll also have to add it to the rest of this file. thought that to be out of scope.

your call